08-graphs/01-graphs.md

```diff
@@ -169,12 +319,28 @@ adjacency_list_with_dict = {
     }
 ```
 
+This representation provides similar benefits to using an array for an adjacency list, but provides us more flexibility with keys.  We no longer need numeric IDs for our nodes and the IDs do not need to be consecutive.  For example, in the adjacency matrix and list above, there was no node 0, but we maintained space for a node with that id.
 
-![Adjacency list diagram with a dictionary](images/graphs-adjacency-list-cities.png)
+#### Adjacency List Big O
+Generally adjacency lists give a good balance between time and space complexity and perform relatively well. 
+- **O(d)** lookup time to find out if any two nodes share an edge, where _d_ is the _degree_ of a node meaning the number of edges the node has
+  - The maximum value of d would be the case where a node is connected to every other node (assuming no duplicate edges) which is _N_ - 1 where _N_ is the total number of nodes in the graph.
+  - When _d_ begins to approach _N_ for the majority of nodes - in other words, when we have a dense graph - we start to see the advantages of using an adjacency matrix with O(1) lookup time over an adjacency list
+- O(2E) = **O(E)** space complexity to store this representation because each element of the array stores a list of the edges a node has
+  -  Each edge is stored twice in the case of an undirected graph
+- **O(1)** time complexity to get all the neighbors of a particular node because it is an O(1) operation to lookup a node by its index and the list at that location is a list of all the node's neighbors.
+
+## Summary
+
+A graph is a data structure consisting of a finite collection of _nodes_, also called _vertices_, and a collection of connections between nodes known as _edges_.  A graph can be either a _directed graph_, or an _undirected graph_.  In a directed graph, edges are not bidirectional: we may be able to travel from node A to node B, but not from node B to node A. Graphs can also have _weighted_ or _unweighted_ edges. Weighteds graphs have some cost or weight assigned to each edge, making it so that the shortest, most direct path from Node A to Node B may not be the cheapest.
```

Comment:
```suggestion
A graph is a data structure consisting of a finite collection of _nodes_, also called _vertices_, and a collection of connections between nodes known as _edges_.  A graph can be either a _directed graph_, or an _undirected graph_.  In a directed graph, edges are not bidirectional: we may be able to travel from node A to node B, but not from node B to node A. Graphs can also have _weighted_ or _unweighted_ edges. Weighted graphs have some cost or weight assigned to each edge, making it so that the shortest, most direct path from Node A to Node B may not be the cheapest.
```